model/jpa/src/main/java/org/keycloak/models/jpa/JpaIDPProvider.java

```diff
@@ -324,6 +330,104 @@ public long count() {
+
+        return model;
+    }
+
+    @Override
+    public void updateMapper(IdentityProviderMapperModel model) {
+        IdentityProviderMapperEntity entity = getMapperEntityById(model.getId(), true);
+        if (!model.getName().equals(entity.getName())) {
+            checkUniqueMapperNamePerIdentityProvider(model);
+        }
+
+        entity.setName(model.getName());
+        entity.setIdentityProviderAlias(model.getIdentityProviderAlias());
+        entity.setIdentityProviderMapper(model.getIdentityProviderMapper());
+        entity.setConfig(model.getConfig());
+    }
+
+    @Override
+    public boolean removeMapper(IdentityProviderMapperModel model) {
+        em.remove(getMapperEntityById(model.getId(), true));
+        return true;
+    }
+
+    @Override
+    public void removeAllMappers() {
+        CriteriaBuilder builder = em.getCriteriaBuilder();
+        CriteriaDelete<IdentityProviderMapperEntity> delete = builder.createCriteriaDelete(IdentityProviderMapperEntity.class);
+        Root<IdentityProviderMapperEntity> mapper = delete.from(IdentityProviderMapperEntity.class);
+        delete.where(builder.equal(mapper.get("realmId"), getRealm().getId()));
+        em.createQuery(delete).executeUpdate();
+    }
+
+    @Override
+    public IdentityProviderMapperModel getMapperById(String id) {
+        return toModel(getMapperEntityById(id, false));
+    }
+
+    @Override
+    public IdentityProviderMapperModel getMapperByName(String identityProviderAlias, String name) {
+        CriteriaBuilder builder = em.getCriteriaBuilder();
+        CriteriaQuery<IdentityProviderMapperEntity> query = builder.createQuery(IdentityProviderMapperEntity.class);
+        Root<IdentityProviderMapperEntity> mapper = query.from(IdentityProviderMapperEntity.class);
+
+        Predicate predicate = builder.and(
+                builder.equal(mapper.get("realmId"), getRealm().getId()),
+                builder.equal(mapper.get("identityProviderAlias"), identityProviderAlias),
+                builder.equal(mapper.get("name"), name));
+
+        TypedQuery<IdentityProviderMapperEntity> typedQuery = em.createQuery(query.select(mapper).where(predicate));
+        try {
+            return toModel(typedQuery.getSingleResult());
+        } catch (NoResultException nre) {
+            return null;
+        }
+    }
+
+    @Override
+    public Stream<IdentityProviderMapperModel> getMappersStream() {
+        CriteriaBuilder builder = em.getCriteriaBuilder();
+        CriteriaQuery<IdentityProviderMapperEntity> query = builder.createQuery(IdentityProviderMapperEntity.class);
+        Root<IdentityProviderMapperEntity> mapper = query.from(IdentityProviderMapperEntity.class);
+
+        Predicate predicate = builder.equal(mapper.get("realmId"), getRealm().getId());
+
+        TypedQuery<IdentityProviderMapperEntity> typedQuery = em.createQuery(query.select(mapper).where(predicate));
+
+        return typedQuery.getResultStream().map(this::toModel);
```

Comment:
Don't we need to wrap the stream using `closing`? Same for other query methods.

Reply:
Good point, we probably need it, yes

Reply:
Will add it, thanks